rfc/bulksave.md

```diff
@@ -0,0 +1,220 @@
+
+## Design
+
+1. Deck calls the gate api:
+
+   1.1	url: 		http://<ip address>:8084/pipelines/bulksave
+   method: 	POST
+   payload:	[{"keepWaitingPipelines":false,"limitConcurrent":true,"application":"apptest","spelEvaluator":"v4",
+   "lastModifiedBy":"admin","name":"pipeline_test","stages":[{"refId":"1","requisiteStageRefIds":[],
+   "type":"wait","name":"Wait","waitTime":30}],"index":3,"id":"66a5c77b-3998-495d-ad61-117313c5800c",
+   "triggers":[],"updateTs":"1599652266000"},{"keepWaitingPipelines":false,"limitConcurrent":true,
+   "application":"apptest","spelEvaluator":"v4","lastModifiedBy":"admin","name":"pipeline_test1",
+   "stages":[{"refId":"1","requisiteStageRefIds":[],"type":"wait","name":"Wait","waitTime":30}],
+   "index":4,"id":"66a5c77b-3998-495d-ad61-117313c5800c","triggers":[],"updateTs":"1599652266000"}......]
+
+
+2. Gate api flow:
+
+   2.1	The post request sent by the deck will hit the PipelineController Rest Controller endpoint.
+
+        com.netflix.spinnaker.gate.controllers.PipelineController.bulksavePipeline(Body pipeline..)
+   2.2 	TaskService will create a new task(synchronous job) for saving the pipeline.
+        taskService.createAndWaitForCompletion(operation)
+
+   2.3	The above new task will call the orca service post api to save the pipeline.
+
+        orcaService.doOperation(@Body Map<String, ? extends Object> body)
+        url: 		http://<ip address>:8083/ops
+        method: 	POST
+
+3.  Orca api flow:
+
+    3.1	The post request sent by the gate will hit the OperationsController Rest Controller endpoint.
+
+        com.netflix.spinnaker.orca.controllers.OperationsController Map<String, String> ops(@RequestBody Map input) {
+
+    3.2	The above method will start a new task with QueueExecutionRunner.kt.
+    3.3	The QueueExecutionRunner will call the StartExecutionHandler.kt.
+    3.4	The StartExecutionHandler.kt will call the SavePipelineTask.java.
+    3.5	SavePipelineTask will call the front 50 rest api.
+        Response response = front50Service.savePipelineList(List<pipeline>);
+    	url: 		http://<ip address>:8080/pipelines/bulksave
+    	method: 	POST
+
+4.  Fiat api flow:
+
+    4.1	Front 50 will check with fiat for permissions.
+
+        @PreAuthorize( "@fiatPermissionEvaluator.storeWholePermission() and
+        authorizationSupport.hasRunAsUserPermission(List<pipeline>)")
+
+    4.2 storeWholePermission() will check if the permissions object is not null..
+    4.3	hasPermission() will check whether that particular resource(APPLICATION) has ‘WRITE’ permission to save or not.
+    4.4	authorizationSupport.hasRunAsUserPermission() check happens at front50,
+        which calls fiat to check whether that particular resource(Application) has ‘EXECUTE’ permission or not.
+
+5.  Front50 api flow:
+
+    5.1	The post request sent by the orca will hit the PipelineController Rest Controller endpoint.
+
+        com.netflix.spinnaker.front50.controllers.OperationsController.bulksave(
```

Comment:
I don't know what `OperationsController` is, I can only assume it's actually going to call the `PipelineController.batchUpdate()` endpoint and we'll be handling whether it's admin vs. non-admin?

Reply:
Caveat that the `batchUpdate()` endpoint wasn't expected to be exposed and lacks the validation provided by `save()`.

Reply:
It should be PipelineController and not OperationsController. Corrected the document too.